sycl/include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp

```diff
@@ -0,0 +1,283 @@
+  annotated_arg(const annotated_arg &) = default;
+  annotated_arg &operator=(annotated_arg &) = default;
+
+  annotated_arg(const T &_obj,
+                const property_list_t &PropList = properties{}) noexcept
+      : obj(_obj) {}
+
+  template <typename... PropertyValueTs>
+  annotated_arg(const T &_obj, PropertyValueTs... props) noexcept : obj(_obj) {
+    static_assert(
+        std::is_same<property_list_t,
+                     detail::merged_properties_t<
+                         property_list_t,
+                         detail::properties_t<PropertyValueTs...>>>::value,
+        "The property list must contain all properties of the input of the "
+        "constructor");
+  }
+
+  // Constructs an annotated_arg object from another annotated_arg object.
+  // The property set PropertyListT contains all properties of the input
+  // annotated_arg object. If there are duplicate properties present in the
+  // property list of the input annotated_arg object, the values of the
+  // duplicate properties must be the same.
+  template <typename T2, typename PropertyList2>
+  explicit annotated_arg(const annotated_arg<T2, PropertyList2> &other) noexcept
+      : obj(other.obj) {
+    static_assert(std::is_convertible<T2, T>::value,
+                  "The underlying data type of the input annotated_arg is not "
+                  "compatible");
+
+    static_assert(
+        std::is_same<
+            property_list_t,
+            detail::merged_properties_t<property_list_t, PropertyList2>>::value,
+        "The constructed annotated_arg type must contain all the properties of "
+        "the input annotated_arg");
+  }
+
+  template <typename T2, typename PropertyListU, typename PropertyListV>
+  explicit annotated_arg(const annotated_arg<T2, PropertyListU> &other,
+                         const PropertyListV &proplist) noexcept
+      : obj(other.obj) {
+    static_assert(std::is_convertible<T2, T>::value,
+                  "The underlying data type of the input annotated_arg is not "
+                  "compatible");
+
+    static_assert(
+        std::is_same<property_list_t, detail::merged_properties_t<
+                                          PropertyListU, PropertyListV>>::value,
+        "The property list of constructed annotated_arg type must be the union "
+        "of the input property lists");
+  }
+
+  operator T() noexcept { return obj; }
+  operator const T() const noexcept { return obj; }
+
+  template <class RelayT = T>
+  std::enable_if_t<detail::HasSubscriptOperator<RelayT>::value,
+                   const decltype(std::declval<RelayT>().operator[](0))> &
+  operator[](std::ptrdiff_t idx) const noexcept {
+    return obj.operator[](idx);
+  }
+
+  template <class RelayT = T>
+  std::enable_if_t<detail::HasSubscriptOperator<RelayT>::value,
+                   decltype(std::declval<RelayT>().operator[](0))> &
+  operator[](std::ptrdiff_t idx) noexcept {
+    return obj.operator[](idx);
+  }
```

Comment:
Non-`const` overload of `operator[]` is not documented by the spec

Reply:
Remove this overload